packages/snaps-utils/src/url.ts

```diff
@@ -0,0 +1,97 @@
+import { type SnapId } from '@metamask/snaps-sdk';
+import { type Infer, is, pattern, string } from '@metamask/superstruct';
+import { assert } from '@metamask/utils';
+
+import { assertIsValidSnapId, stripSnapPrefix } from './snaps';
+
+export const METAMASK_URL_REGEX =
+  /^metamask:\/\/(?<authority>(client|snap))(?<path>.+)$/u;
+
+export const MetaMaskUrlStruct = pattern(string(), METAMASK_URL_REGEX);
+
+export type MetaMaskUrl = Infer<typeof MetaMaskUrlStruct>;
+
+export type Authority = 'client' | 'snap';
+
+export const ExtensionPaths = ['/'];
+
+export const SnapPaths = ['/home'];
+
+/**
+ * Parse a url string to an object containing the authority and path.
+ * This validates the url before parsing it.
+ *
+ * @param str - The url string to validate and parse.
+ * @returns The parsed url.
+ * @throws If the authority or path is invalid.
+ */
+export function parseMetaMaskUrl(str: MetaMaskUrl): {
+  authority: Authority;
+  snapId?: SnapId;
+  path: string;
+} {
+  const baseErrorMessage = 'Invalid MetaMask url:';
+  try {
+    const url = new URL(str);
+    const { hostname: authority, pathname: path, protocol } = url;
+    if (protocol !== 'metamask:') {
+      throw new Error('invalid scheme.');
+    }
+    if (authority === 'client') {
+      assert(
+        ExtensionPaths.includes(path),
+        `${baseErrorMessage} invalid client path.`,
+      );
+      return {
+        authority,
+        path,
+      };
+    } else if (authority === 'snap') {
+      const strippedPath = stripSnapPrefix(path.slice(1));
+      const location = path.slice(1).startsWith('npm:') ? 'npm:' : 'local:';
+      const isNameSpaced = strippedPath.startsWith('@');
```

Comment:
Wouldn't it be easier to just get the snap ID and path, and do the assertions based on that?

```ts
/**
 * Parse a Snap URL into its `snapId` and `path` segments.
 *
 * @param path - The Snap URL to parse.
 * @returns The `snapId` and `path` segments.
 * @example
 * const path = '@foo/bar/baz';
 * const { snapId, path } = parseSnapUrl(path);
 *
 * console.log(snapId); // '@foo/bar'
 * console.log(path); // 'baz'
 * @example
 * const path = 'bar/baz';
 * const { snapId, path } = parseSnapUrl(path);
 *
 * console.log(snapId); // 'bar'
 * console.log(path); // 'baz'
 */
function parseSnapUrl(path: string) {
  if (path.startsWith('@')) {
    const [scope, name, ...url] = path.split('/');
    assert(scope, 'Expected `scope` to be defined.');
    assert(name, 'Expected `name` to be defined.');

    return {
      snapId: `${scope}/${name}`,
      path: url.join('/'),
    };
  }

  const [name, ...url] = path.split('/');
  assert(name, 'Expected `name` to be defined.');

  return {
    snapId: name,
    path: url.join('/'),
  };
}

const { snapId, path } = parseSnapUrl(path);
// ...
```

Reply:
This doesn't work for local snaps

Reply:
Good point, but how do we possibly know where the URL of a local Snap ends, and the path to navigate to starts?

Reply:
A valid path portion of a url of a local snap will always be of the same structure `/local:https://localhost:someport/path`. So if the snap id is valid then the split should always have 4 members and then checking the last member should always be one of the paths we allow, which currently is just home.

Reply:
What if the local Snap has a path in the URL?

Reply:
What do you mean? If it's something like `local:https://localhost:8080/foo/home` then that would throw as an invalid path because the split would result in 5 members

Reply:
But shouldn't that be valid? 🤔

Reply:
No, how come? I mean technically yeah it's valid, but not in our world haha -- I thought we aren't allowing nested paths (yet) and just from the white list of snap paths?

Reply:
I mean, `local:https://localhost:8080/foo` is a valid _snap ID_.

Reply:
Ah didn't know we allowed that, well that changes things.

Reply:
[eaa563b](https://github.com/MetaMask/snaps/pull/2719/commits/eaa563b0129c8c7a5f2b946ca5cd849b663a3dbf)